### PR TITLE
Add `BetaLiIpProfile` to free boundary equilibrium solver

### LINKED ISSUE #1042: `BetaIpLiProfile` based on a flux function parameterisation and plasma integral matching
## Description of issue / requirement to address

With the refactor of the iterators #1027, we'll be losing some functionality; namely the ability to constrain the plasma profile `l_i` integral during iteration. I'm not sure yet if this actually helps convergence, to be honest, and it may be that we need to re-instate this functionality down the line.

With `BetaIpProfile` one can constrain `beta_p` but not `l_i` simultaneously.

We should however have the functionality to let the user determine their own flux functions based on matching the desired plasma integral values (`beta_p` and `l_i`) for a given plasma shape and profile parameterisation. 

## Proposed solution

Add `BetaIpLiProfile` which provides flux function optimisation preserving `beta_p` `l_i` and `I_p`.

## Alternative solutions

We're mostly moving towards using `CustomProfile` with pre-ascertained flux functions when solving equilibria nowadays. We sacrifice consistency on plasma integrals slightly, but provided the flux functions come from a transport solver of some kind, it is generally felt this is acceptable. The integrals `beta_p` and `l_i` are effectively results. This is fine, but requires a user to have flux functions that provide the desired integrals. As these flux functions in any case tend to come from transport solvers with a fixed boundary and no X-point(s) the integrals will only ever be approximate.

### PR DESCRIPTION
**Pull request by CoronelBuendia**
## Linked Issues

Closes #1042

## Description

Adds a `BetaLiIpProfile` to free boundary equilibrium solver, which is just resurrected stuff from BLUEPRINT. Turns out it works pretty great for EU-DEMO, at least for the example considered. It's not the most beautiful implementation, and uses scipy optimize at present, which I hate with a passion, but... it works out of the box.
## Checklist

I confirm that I have completed the following checks:

- [x] Tests run locally and pass `pytest tests --reactor`
- [x] Code quality checks run locally and pass `flake8` and `black .`
- [x] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

**Comment by CoronelBuendia** [comment]
going to add tests

**Review by je-cook — COMMENTED** [review]
Looks ok, just one comment

File: tests/equilibria/setup_methods.py [hunk 1]
```diff
@@ -86,8 +86,10 @@ def _coilset_setup(self, materials=False):
 
     if materials:
         # Max PF currents / sizes don't stack up in the CREATE document...
-        self.coilset.assign_coil_materials("PF", j_max=12.5, b_max=12.5)
-        self.coilset.assign_coil_materials("CS", j_max=12.5, b_max=12.5)
+        self.coilset.get_coiltype("PF").b_max = 12.5
+        self.coilset.get_coiltype("PF").j_max = 12.5e6
+        self.coilset.get_coiltype("CS").b_max = 12.5
+        self.coilset.get_coiltype("CS").j_max = 12.5e6
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply tests/equilibria/setup_methods.py:92]: So `CoilSet.assign_coil_materials` doesn't exist any more, but I have just discovered `CoilSet.assign_material`, will update now

[the remark this answers is not in the document]
Comment (je-cook) [reply tests/equilibria/setup_methods.py]: ahh sorry my fault!

**Review by je-cook — APPROVED** [review]

**Pull request merged**